File: app/src/main/java/ru/magflayer/spectrum/domain/interactor/ColorPhotoInteractor.java
package ru.magflayer.spectrum.domain.interactor;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import ru.magflayer.spectrum.domain.entity.ColorPhotoEntity;
import ru.magflayer.spectrum.domain.repository.PhotoRepository;
import rx.Observable;

public class ColorPhotoInteractor {

    private final PhotoRepository photoRepository;

    @Inject
    ColorPhotoInteractor(final PhotoRepository photoRepository) {
        this.photoRepository = photoRepository;
    }

    public Observable<Boolean> saveColorPhoto(final ColorPhotoEntity entity) {
        return photoRepository.savePhoto(entity);
    }

    public Observable<List<ColorPhotoEntity>> loadColorPhotos() {
        return photoRepository.loadPhotos()
                .map(entities -> {
                    List<ColorPhotoEntity> photoEntities = new ArrayList<>();

                    for (ColorPhotoEntity entity : entities) {
                        if (new File(entity.getFilePath()).exists()) {
                            photoEntities.add(entity);
                        } else {
                            photoRepository.removePhoto(entity);
                        }
                    }

                    return photoEntities;
                });
    }

    public Observable<Boolean> removeColorPhoto(final ColorPhotoEntity entity) {
        return photoRepository.removePhoto(entity)
                .map(success -> new File(entity.getFilePath()).delete());
    }

    public Observable<ColorPhotoEntity> loadColorPhoto(final String filePath) {
        return photoRepository.loadPhoto(filePath);
    }

}

File: app/src/main/java/ru/magflayer/spectrum/domain/interactor/ColorPhotoInteractor.kt
package ru.magflayer.spectrum.domain.interactor

import ru.magflayer.spectrum.domain.entity.ColorPhotoEntity
import ru.magflayer.spectrum.domain.repository.FileManagerRepository
import ru.magflayer.spectrum.domain.repository.PhotoRepository
import java.io.File
import javax.inject.Inject

class ColorPhotoInteractor @Inject constructor(
    private val photoRepository: PhotoRepository,
    private val fileManagerRepository: FileManagerRepository
) {

    suspend fun saveColorPhoto(entity: ColorPhotoEntity): Boolean {
        return photoRepository.savePhoto(entity)
    }

    suspend fun loadColorPhotos(): List<ColorPhotoEntity> {
        return photoRepository.loadPhotos()
            .map { colorPhoto ->
                colorPhoto.also {
                    val fileExists = fileManagerRepository.isFileExists(colorPhoto.filePath)
                    if (!fileExists) {
                        photoRepository.removePhoto(colorPhoto)
                    }
                }
            }
            .sortedByDescending { colorPhoto -> colorPhoto.millis }
    }

    suspend fun removeColorPhoto(entity: ColorPhotoEntity): Boolean {
        val success = photoRepository.removePhoto(entity)
        if (entity.type === ColorPhotoEntity.Type.INTERNAL) {
            File(entity.filePath).delete()
        }
        return success
    }

    suspend fun loadColorPhoto(filePath: String): ColorPhotoEntity {
        return photoRepository.loadPhoto(filePath)
    }

}
